proposals/gc/MVP-JS.md

```diff
@@ -0,0 +1,532 @@
+instantiation-time and run-time checks.
+
+But what about the other direction: what if WebAssembly exports a struct
+type definition that uses a `ref` field type? To reflect WebAssembly structural
+subtyping, we need new Value Type objects. Since the set of Value Type objects
+is open-ended, the WebAssembly JS API adds a new Value Type constructor
+function: `WebAssembly.ref(T)`. Using this, JS can create a suitable Type
+Definition object whose instances can be passed to the above `example3.wasm`
+because its fields are structurally typed:
+
+```js
+function makeTypes() {
+    const Point = new StructType([{name: "x", type: int32}, {name: "y", type: int32}]);
+    const PointRef = WebAssembly.ref(Point);
+    const Rect = new StructType([{name: "p1", type: PointRef}, {name: "p2", type: PointRef}]);
+    return {Point, Rect};
+}
+
+WebAssembly.instantiateStreaming(fetch('example3.wasm'))
+.then(({instance}) => {
+    const {Point, Rect} = makeTypes();
+    var rect = new Rect(new Point(1,2), new Point(3,4));
+    assert("x" in rect.x1);
+    instance.exports.goWild(rect);
+    assert(!("x" in rect.x1));
+    assert(rect.x1[0] === 10);
+    assert(rect.x1[1] === 20);
+
+    const {Point2:Point, Rect2:Rect} = makeTypes();
+    var rect2 = new Rect2(new Point2(1,2), new Point2(3,4));
+    instance.exports.goWild(rect2);  // ok: structurally equivalent
+});
+```
+
+Having a distinction between these two kinds of subtyping is unfortunate, but
+the expectation is that practical uses will generate all these Type Definition
+objects as part of the usual JS boilerplate glue code and end users would simply
+interact with objects that were either tuple-y or had property names as expected
+from the source language. (For example, a WebAssembly-friendly [Flow](https://flow.org)
+variant might use structural typing for fields with [Tuple Type](https://flow.org/en/docs/types/tuples/)
+and nominal typing for fields with [Class Type](https://flow.org/en/docs/types/classes/).)
+
+In addition to the dynamic checks necessary when JS passes a value to
+WebAssembly (either as an argument to a wasm exported function or a return value
+from a wasm imported function), the instantiation algorithm must ensure that
+imported type definitions' fields' types are compatible. For example, given the
+following module:
+
+```wat
+;; example5.wat --> example5.wasm
+(module
+    (type $Point (import "" "Point") (eq (struct (field i32) (field i32))))
+    (type $Rect (import "" "Rect") (eq (struct (field (ref $Point)) (field (ref $Point)))))
+    ;; ...
+)
+```
+
+this instantiation would succeed:
+
+```js
+const Point = new StructType([{type: int32}, {type: int32}]);
+const Rect = new StructType([{type: Point.ref}, {type: Point.ref}]);
+
+WebAssembly.instantiateStreaming(fetch('example5.wasm'), {'':{Point, Rect}});
+```
+
+while this instantiation would fail because of the `Point1`/`Point2` mismatch in the
```

Comment:
One interesting question here is how these additional link-time type checks are integrated into the semantics. Currently, all link-time checking is defined by the core spec. IIUC, this semantics would require the JS API to define an additional layer of checking that is a refinement of that. Is that correct?

Reply:
Correct; the JS API would have to "traverse" the types of signatures and make sure the type imports' identities all line up.  I'll add a note.